packages/runtime/container-runtime/src/compatUtils.ts

```diff
@@ -0,0 +1,194 @@
+ * TODO: This can be removed after 3.0 is released.
+ */
+const defaultConfigsForPreFF3: IContainerRuntimeOptionsVersionDependent = {
+	gcOptions: {},
+	flushMode: FlushMode.TurnBased,
+	compressionOptions: enabledCompressionConfig,
+	enableRuntimeIdCompressor: undefined as unknown as "on" | "delayed",
+	enableGroupedBatching: true,
+	explicitSchemaControl: false,
+};
+
+/**
+ * Subset of the IContainerRuntimeOptionsInternal properties which are version-dependent.
+ *
+ * @remarks
+ * When a new option is added to IContainerRuntimeOptionsInternal, we must consider if it's a version-dependent option.
+ * If it's considered version-dependent, then a corresponding entry must be added to `versionDependentOptionConfigs`. If not, then
+ * it must be omitted from this type.
+ *
+ * Note: We use `Omit` instead of `Pick` to ensure that all new options are included in this type by default. If any new properties
+ * are added to IContainerRuntimeOptionsInternal, they will be included in this type unless explicitly omitted. This will prevent
+ * us from forgetting to account for any new properties in the future.
+ */
+export type IContainerRuntimeOptionsVersionDependent = Required<
+	Omit<
+		IContainerRuntimeOptionsInternal,
+		| "chunkSizeInBytes"
+		| "maxBatchSizeInBytes"
+		| "loadSequenceNumberVerification"
+		| "summaryOptions"
+		| "compatibilityMode"
+	>
+>;
+
+/**
+ * Default configurations for version-dependent options based on the compatibilityMode.
+ */
+interface IVersionDependentOptionConfig<
+	K extends keyof IContainerRuntimeOptionsVersionDependent,
+> {
+	/**
+	 * The minimum version of the container runtime that is required to use the modern config for the
+	 * option. This will compared with the compatibilityMode to determine if the modern or legacy config
+	 * should be used by default.
+	 * This should be undefined if no versions are ready to use the modern config by default. In this case,
+	 * we will always default to the legacy config.
+	 */
+	minVersionForModernConfig: string | undefined;
+	/**
+	 * The default config for the version-dependent option that ensures clients that do not understand the
+	 * modern config will not break.
+	 * This is the config that will be used by default when the compatibilityMode is less than minVersionForModernConfig.
+	 */
+	legacyConfig: IContainerRuntimeOptionsVersionDependent[K];
+	/**
+	 * The default config for the version-dependent option when all clients are expected to understand the modern config.
+	 * This is the config that will be used by default when the compatibilityMode is at least minVersionForModernConfig.
+	 */
+	modernConfig: IContainerRuntimeOptionsVersionDependent[K];
+}
+
+/**
+ * Mapping of version-dependent options to their compatibility related configs.
+ */
+const versionDependentOptionConfigMap: {
+	[K in keyof IContainerRuntimeOptionsVersionDependent]: IVersionDependentOptionConfig<K>;
+} = {
+	enableGroupedBatching: {
```

Comment:
i think this structure would be better if the values were also a map of version to default:
```
enableGroupedBatching: {
"1.0.0": false,
"2.0.0": true,
}
```
then given a version, you build the defaults based on the value with the highest matching semver.

this will make it easier to add something like "2.5.0", as it is just a new entry. i believe this also get rid of the legacy vs modern concept.

Reply:
@anthony-murphy, I like this idea. I tried it out in the latest commit